### ISSUE TITLE
Bug: Code coverage is not calculated when using Vitest with Rule Tester

### ISSUE THREAD

**Issue by azat-io** [typescript-eslint/typescript-eslint#7241]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

rule-tester

### Repro Code

I use [Vitest](https://vitest.dev).

I tried to upgrade `typescript-eslint` to version 6 in my project and encountered some problems integrating the `@typescript-eslint/rule-tester` package.

In the beginning, I was getting an error:

```
Error: Missing definition for `afterAll` - you must set one using `RuleTester.afterAll` or there must be one defined globally as `afterAll`.
```

I was able to fix this problem in the following way:

```ts
import { RuleTester } from '@typescript-eslint/rule-tester'
import { afterAll, describe, it } from 'vitest'

describe(RULE_NAME, () => {
  RuleTester.afterAll = afterAll
  RuleTester.describe = describe

  let ruleTester = new RuleTester({
    parser: '@typescript-eslint/parser',
  })

  // ....
})
```

If this is the correct way, you should add instructions on how to integrate the rule-tester package with Vitest.

At the moment I have a different problem. The code coverage has stopped being calculated.

Can you tell me how I can fix this problem? 

Here is PR in my project:
https://github.com/azat-io/eslint-plugin-perfectionist/pull/37

Steps to reproduce:

1. Clone [eslint-plugin-perfectionist](https://github.com/azat-io/eslint-plugin-perfectionist) repo
2. Switch to `feat/typescript-eslint-6` branch
3. Run `pnpm install`
4. Run `pnpm test:unit`

### Expected Result

Code coverage is calculated correctly.

### Actual Result

Code coverage is 0.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/rule-tester`        | `6.0.0` |
| `@typescript-eslint/utils`             | `6.0.0` |
| `@vitest/coverage-v8`                           | `0.33.0` |
| `vitest`                               | `0.33.0` |

**Comment by bradzacher** [comment]
For good measure you should also set `RuleTester.it`.

The only difference between v5 and v6 is that in v5 our rule tester was a subclass of eslint's - and in v6 our rule tester is a fork. However the underlying infra is exactly the same. The only reason we forked was to make it easier to add our custom logic.

This means that the integration with vite is should be the same as it was before - line coverage should not have changed.

However I don't have any experience with vite so I don't have any good advice to help you with this. @JoshuaKGoldberg youve used vut haven't you?

**Comment by azat-io** [comment]
Yes, I added `RuleTester.it`, but that doesn't solve the problem of test coverage not working correctly.

**Comment by JoshuaKGoldberg** [comment]
😞 I truly have no idea why this is happening. Looks like a bug that needs investigation.

**Comment by azat-io** [comment]
Have you tried investigating the reason for this behavior?

**Comment by JoshuaKGoldberg** [comment]
I haven't had time and we probably won't very a long while. It's all yours if you're up for it!

**Comment by bradzacher** [comment]
Given the bandwidth constraints of us maintainers - it's unlikely we'll get around to this ourselves any time soon. We don't use vitest so we would have to learn it, setup an environment, test, then debug - it's a large time investment!

**Comment by u3u** [comment]
Work normally with me, refer to:

https://app.codecov.io/gh/u3u/eslint-plugin-arrow-return-style/tree/main/src%2Frules
https://github.com/u3u/eslint-plugin-arrow-return-style/blob/main/src/rules/arrow-return-style.test.ts

```ts
import { RuleTester } from '@typescript-eslint/rule-tester';
import dedent from 'dedent';
import { afterAll, describe, it } from 'vitest';
import { arrowReturnStyleRule, RULE_NAME } from './arrow-return-style';

RuleTester.afterAll = afterAll;
RuleTester.describe = describe;
RuleTester.it = it;

const ruleTester = new RuleTester({
  parser: '@typescript-eslint/parser',

  parserOptions: {
    ecmaFeatures: { jsx: true },
  },
});

ruleTester.run(RULE_NAME, arrowReturnStyleRule, {
  invalid: [],

  valid: [],
});
```

**Comment by azat-io** [comment]
@u3u Thank you for your example!

Moving the tests outside of the `it` function fixes this problem.

The problem is in the `it` function.

**Comment by azat-io** [comment]
@bradzacher I think it's a bug. We can't use the `it` function from Vitest. Could you please reopen the issue?

**Comment by azat-io** [comment]
The problem is somewhere in the `it` function.

Counts coverage correctly:

```ts
RuleTester.afterAll = afterAll;
RuleTester.describe = describe;
RuleTester.it = it;

const ruleTester = new RuleTester({
  parser: '@typescript-eslint/parser',
});

describe('test', () => {
  ruleTester.run(RULE_NAME, arrowReturnStyleRule, {
    invalid: [ /* ... */ ],
    valid: [ /* ... */ ],
  });
});
```

Doesn't count coverage:

```ts
RuleTester.afterAll = afterAll;
RuleTester.describe = describe;
RuleTester.it = it;

const ruleTester = new RuleTester({
  parser: '@typescript-eslint/parser',
});

describe('test', () => {
  it('test', () => {
    ruleTester.run(RULE_NAME, arrowReturnStyleRule, {
      invalid: [ /* ... */ ],
      valid: [ /* ... */ ],
    });
  })
});
```

@bradzacher @JoshuaKGoldberg Please reopen this issue. 🙏

**Comment by bradzacher** [comment]
I'd say things get a bit funky when you nest the rule tester within an `it` because it already defines a `describe` and `it`

So by wrapping it you're essentially doing

```ts
describe('test', () => {
  it('test', () => {
    describe(RULE_NAME, () => {
      describe('valid', () => {
        it('valid_test_one', () => {...});
        it('valid_test_two', () => {...});
        //...
      });
    }); 
  });
});
```

So I'm surprised that vite doesn't error out when you nest a describe within an it (jest errors on that structure iirc)
